org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/XMLWorkspaceService.java

```diff
@@ -44,9 +45,15 @@ public void didChangeConfiguration(DidChangeConfigurationParams params) {
 
 	@Override
 	public void didChangeWatchedFiles(DidChangeWatchedFilesParams params) {
-
+		saveChangeWatchedFiles(params);
 	}
 
-	
+	private void saveChangeWatchedFiles(DidChangeWatchedFilesParams params) {
+		XMLTextDocumentService xmlTextDocumentService = (XMLTextDocumentService) xmlLanguageServer.getTextDocumentService();
+		List<FileEvent> changes = params.getChanges();
+		for (FileEvent change: changes) {

```

Comment:
if the file is already opened, i.e. tracked by the server, then it should not be revalidated, or else the job will be done twice

Reply:
> if the file is already opened,

Are you referring to the xml file or xsd file? In my test case, the xsd file is not opened while the xml is. I'm not sure when the validation would be done twice, because, without this code snippet, validation for opened xmls won't happen at all when xsd files (open or unopen) are changed externally. Could you explain a bit more about when validation might happen twice?

Reply:
if the xsd is also opened in lsp4xml, saving it would trigger both documentChange and fileWatchChanged events. But only the document change is interesting in that case